### PR TITLE
Galera cache encryption

### PR DESCRIPTION
**Pull request by mkaruza**
* Setting encryption key to provider
* Implementation of do_crypt function which is used for enc / dec
* Updated submodule

### CODE REVIEW

File: sql/wsrep_server_service.cc [hunk 1]
```diff
@@ -293,3 +293,48 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+                           (unsigned char*)iv, encryption_service::iv_size,
+                           mode | ENCRYPTION_FLAG_NOPAD,
+                           encryption_service::key_id,
+                           encryption_service::key_version))
+    {
+      throw wsrep::runtime_error("Failed encryption_ctx_init()");
+    }
+  }
+  unsigned int ctx_update_size= 0;
+  if(encryption_ctx_update(*ctx, (const unsigned char*)input.data(), (unsigned int)input.size(),
```
[the remark this answers is not in the document]
Comment (mkaruza) [reply sql/wsrep_server_service.cc]: OK

File: sql/wsrep_server_service.cc [hunk 2]
```diff
@@ -293,3 +293,48 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+    throw wsrep::runtime_error("Failed encryption_ctx_update()");
+  }
+  unsigned int ctx_finish_size= 0;
+  if (final)
+  {
+    if(encryption_ctx_finish(*ctx, (unsigned char *)output + ctx_update_size,
+                              &ctx_finish_size))
+    {
+      throw wsrep::runtime_error("Failed encryption_ctx_finish()");
+    }
```
[the remark this answers is not in the document]
Comment (mkaruza) [reply sql/wsrep_server_service.cc]: Added

File: sql/wsrep_server_service.cc [hunk 3]
```diff
@@ -293,3 +293,48 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+                                    void*                 output,
+                                    bool                  encrypt,
+                                    bool                  final)
+{
+  int mode= encrypt ? ENCRYPTION_FLAG_ENCRYPT : ENCRYPTION_FLAG_DECRYPT;
+
+  if (*ctx == NULL)
+  {
+    *ctx = ::malloc(encryption_ctx_size(encryption_service::key_id,
+                                                encryption_service::key_version));
```
[the remark this answers is not in the document]
Comment (mkaruza) [reply sql/wsrep_server_service.cc]: OK

File: sql/wsrep_server_service.h [hunk 4]
```diff
@@ -27,6 +27,17 @@ class Wsrep_server_state;
...
+
+private:
+  
+  struct encryption_service
+  {
+    static const size_t key_version    = 1;
+    static const size_t key_id         = 1;
+    static const size_t iv_size        = 16;
+    static const size_t aes_block_size = 16;
+  };
```
[the remark this answers is not in the document]
Comment (mkaruza) [reply sql/wsrep_server_service.h]: Added const encryption_iv_size into wsrep_server_service.cc

File: sql/wsrep_server_service.cc [hunk 5]
```diff
@@ -293,3 +293,48 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+
+void Wsrep_server_service::do_crypt(void**                ctx,
+                                    wsrep::const_buffer&  key,
+                                    const char            (*iv)[32],
+                                    wsrep::const_buffer&  input,
+                                    void*                 output,
+                                    bool                  encrypt,
+                                    bool                  final)
+{
+  int mode= encrypt ? ENCRYPTION_FLAG_ENCRYPT : ENCRYPTION_FLAG_DECRYPT;
```
[the remark this answers is not in the document]
Comment (mkaruza) [reply sql/wsrep_server_service.cc]: Moved

File: sql/wsrep_server_service.cc [hunk 6]
```diff
@@ -302,3 +302,78 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+                                         const void* key,
+                                         size_t size,
+                                         unsigned int version)
+{
+  output.resize(key_version_preamble_size + size);
+  memcpy(output.data(), (void *)&version, sizeof(version));
+  memcpy(output.data() + key_version_preamble_size, key, size);
+}
+
+void Wsrep_server_service::key_deserialize(void const* input,
```
Comment (ayurchen) [inline sql/wsrep_server_service.cc]: I think you meant `const void*` here ) `void* const` is a totally different thing.

File: sql/wsrep_server_service.cc [hunk 7]
```diff
@@ -302,3 +302,78 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+                                           const size_t& input_size,
```
Comment (ayurchen) [inline sql/wsrep_server_service.cc]: And passing size_t by reference here is perfectly fine, but a bit misleading. In total the signature of this method implies that both contents of `input` buffer and its length can be changed

File: sql/wsrep_server_service.cc [hunk 8]
```diff
@@ -302,3 +302,78 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
 void Wsrep_server_service::debug_sync(const char*)
 {
 }
+
+// IV vector is the size of a block, which for AES is 16
+const unsigned int encryption_iv_size = MY_AES_BLOCK_SIZE;
```
Comment (ayurchen) [inline sql/wsrep_server_service.cc]: This constant seems to be used only once. Why not using `MY_AES_BLOCK_SIZE` directly there? Or, on the contrary, why not use the whole length of supplied IV - apparently `encryption_ctx_init()` allows passing IV of arbitrary size...

File: sql/wsrep_server_service.cc [hunk 9]
```diff
@@ -302,3 +302,78 @@ int Wsrep_server_service::wait_committing_transactions(int timeout)
...
+  memcpy(output.data() + key_version_preamble_size, key, size);
+}
+
+void Wsrep_server_service::key_deserialize(void const* input,
+                                           const size_t& input_size,
+                                           const void*& key_ptr,
+                                           size_t& size,
+                                           unsigned int& version)
+{
+  memcpy((void *)&version, input, sizeof(version));
```
Comment (ayurchen) [inline sql/wsrep_server_service.cc]: I think one needs a runtime check `input_size >= sizeof(version)` before doing that - and I suppose return an error code.

**Pull request closed without merging**